RecoHGCal/TICL/plugins/TracksterLinkingbyGNN.cc

```diff
@@ -0,0 +1,443 @@
+
+        features.push_back(ts.raw_energy());        // 16: raw_energy
+        features.push_back(ts.raw_em_energy());     // 17: raw_em_energy
+
+        features.push_back(ts.id_probabilities(0)); // 18: photon probability
+        features.push_back(ts.id_probabilities(1)); // 19: electron probability
+        features.push_back(ts.id_probabilities(2)); // 20: muon probability
+        features.push_back(ts.id_probabilities(3)); // 21: neutral_pion probability
+        features.push_back(ts.id_probabilities(4)); // 22: charged_hadron probability
+        features.push_back(ts.id_probabilities(5)); // 23: neutral_hadron probability
+
+        features.push_back(min_z_lc.z());           // 24: z_min (minimum z of constituent LCs)
+        features.push_back(max_z_lc.z());           // 25: z_max (maximum z of constituent LCs)
+        features.push_back(length);                 // 26: trackster length from minimum to maximum element in terms of layers
+        // 27 - will be added later in code (if not a gnn) - node degrees
+        // 28 - will be added later in code (if not a gnn) - degrees centrality
+        features.push_back(ts.time());             // 27(29): time
+    }
+
+    std::vector<float> edges_src;
+    std::vector<float> edges_dst;
+    int node_degree_max = 0;
+    
+    for (int i = 0; i < N; i++){
+        const auto &ts_i = tracksters[i];
+
+        std::vector<float> vertex_energy_i;
+        std::vector<float> vertex_i_x;
+        std::vector<float> vertex_i_y;
+        std::vector<float> vertex_i_z;
+
+        const Vector &eigenvector0_i = ts_i.eigenvectors(0);
+
+        for (auto &lc_idx : ts_i.vertices()){
+            auto lc = layerClusters[lc_idx];
+            vertex_energy_i.push_back(lc.energy());
+            vertex_i_x.push_back(lc.x());
+            vertex_i_y.push_back(lc.y());
+            vertex_i_z.push_back(lc.z());
+        }
+        int node_degree = ticlGraph->getNode(i).getInner().size();
+        node_degrees.push_back(node_degree);
+        
+        if (node_degree > node_degree_max){
+          node_degree_max = node_degree;
+        }
+
+        for (auto &i_neighbour : ticlGraph->getNode(i).getInner()){
+
+            const auto &ts_j = tracksters[i_neighbour];
+
+            std::vector<float> vertex_energy_j;
+            std::vector<float> vertex_j_x;
+            std::vector<float> vertex_j_y;
+            std::vector<float> vertex_j_z;
+
+            const Vector &eigenvector0_j = ts_j.eigenvectors(0);
+
+            for (auto &lc_idx : ts_j.vertices()){
+                auto lc = layerClusters[lc_idx];
+                vertex_energy_j.push_back(lc.energy());
+                vertex_j_x.push_back(lc.x());
+                vertex_j_y.push_back(lc.y());
+                vertex_j_z.push_back(lc.z());
+            }
+
+            // Create an edge between the tracksters
+            edges_src.push_back(static_cast<float>(i_neighbour));
```

Comment:
why do we need the cast to float?

Reply:
Because the input to the netowrk has to be float. But I will test if it works without the cast